File: rust/src/bin/aoc_2024_08.rs
// Advent of Code 2024 - Day 8.

use aoc_rust::{coord, Coord};
use itertools::Itertools;
use std::cmp::max;
use std::collections::{HashMap, HashSet};
use std::io;

struct Map {
    height: i32,
    width: i32,
    antenna: HashMap<char, Vec<Coord>>,
}

fn read_map() -> Map {
    let mut height: i32 = 0;
    let mut width: i32 = 0;
    let mut antenna: HashMap<char, Vec<Coord>> = HashMap::new();
    for (line, row) in io::stdin().lines().zip(0..) {
        for (ch, col) in line.unwrap().trim().chars().zip(0..) {
            if ch != '.' {
                antenna
                    .entry(ch)
                    .and_modify(|v| v.push(coord(row, col)))
                    .or_insert(vec![coord(row, col)]);
            }
            height = max(height, row + 1);
            width = max(width, col + 1);
        }
    }
    Map {
        height,
        width,
        antenna,
    }
}

fn check_bounds(m: &Map, coord: Coord) -> Option<Coord> {
    if coord.row >= 0 && coord.row < m.height && coord.col >= 0 && coord.col < m.width {
        Some(coord)
    } else {
        None
    }
}

fn anti_nodes_a(m: &Map) -> HashSet<Coord> {
    let mut locations: HashSet<Coord> = HashSet::new();
    for nodes in m.antenna.values() {
        for v in nodes.iter().combinations(2) {
            let [n1, n2] = v[..] else {
                panic!("Expected 2-vector from combinations(2)")
            };
            let diff = *n2 - *n1;
            check_bounds(m, *n1 - diff).map(|n| locations.insert(n));
            check_bounds(m, *n2 + diff).map(|n| locations.insert(n));
        }
    }
    locations
}

fn anti_nodes_b(m: &Map) -> HashSet<Coord> {
    let mut locations: HashSet<Coord> = HashSet::new();
    for nodes in m.antenna.values() {
        for v in nodes.iter().combinations(2) {
            let [n1, n2] = v[..] else {
                panic!("Expected 2-vector from combinations(2)")
            };
            let diff = *n2 - *n1;
            let mut n = *n1;
            while let Some(c) = check_bounds(m, n) {
                locations.insert(c);
                n += diff;
            }
            n = *n1 - diff;
            while let Some(c) = check_bounds(m, n) {
                locations.insert(c);
                n -= diff;
            }
        }
    }
    locations
}

fn main() {
    let m = read_map();
    println!("{}", anti_nodes_a(&m).len());
    println!("{}", anti_nodes_b(&m).len());
}

File: python/src/aoc_2024_08.py
"""Advent of Code 2024 - Day 8."""

from itertools import combinations
from math import gcd
from sys import stdin

from coord import Coord, Extent


class Map:
    """Puzzle map."""

    def __init__(self, lines: list[str]) -> None:
        height = 0
        width = 0
        self.antenna: dict[str, list[Coord]] = {}
        for y, line in enumerate(lines):
            for x, ch in enumerate(line.strip()):
                height = max(height, y + 1)
                width = max(width, x + 1)
                if ch != ".":
                    if ch in self.antenna:
                        self.antenna[ch].append(Coord(x, y))
                    else:
                        self.antenna[ch] = [Coord(x, y)]
        self.extent = Extent(width, height)

    def print(self, antinodes: set[Coord]) -> None:
        """Print map for debugging."""
        for crd in self.extent.upto_by_y():
            for frequency, locations in self.antenna.items():
                if crd in locations:
                    print(frequency, end="")
                    break
            else:
                if crd in antinodes:
                    print("#", end="")
                else:
                    print(".", end="")
                if crd.x == self.extent.x - 1:
                    print()

    def antinodes(self) -> set[Coord]:
        """Return antinode locations."""
        locations: set[Coord] = set()
        for nodes in self.antenna.values():
            for n1, n2 in combinations(nodes, 2):
                for a in (n1 - (n2 - n1), n2 + (n2 - n1)):
                    if a.in_bounds(self.extent):
                        locations.add(a)
        return locations

    def antinodes_b(self) -> set[Coord]:
        """Return antinode locations for part b."""
        locations: set[Coord] = set()
        for nodes in self.antenna.values():
            for n1, n2 in combinations(nodes, 2):
                d = n2 - n1
                d //= gcd(abs(d.x), abs(d.y))
                c = n1
                while c.in_bounds(self.extent):
                    locations.add(c)
                    c += d
                c = n1 - d
                while c.in_bounds(self.extent):
                    locations.add(c)
                    c -= d
        return locations


if __name__ == "__main__":
    m = Map(stdin.readlines())
    print(len(m.antinodes()))
    print(len(m.antinodes_b()))
